consensus/src/chained_bft/block_storage/pending_votes.rs

```diff
@@ -0,0 +1,175 @@
+    author_to_last_voted_info: HashMap<Author, LastVoteInfo>,
+}
+
+impl PendingVotes {
+    pub fn new() -> Self {
+        PendingVotes {
+            li_digest_to_votes: HashMap::new(),
+            round_to_tc: HashMap::new(),
+            author_to_last_voted_info: HashMap::new(),
+        }
+    }
+
+    pub fn insert_vote(
+        &mut self,
+        vote_msg: &VoteMsg,
+        min_votes_for_qc: usize,
+    ) -> VoteReceptionResult {
+        if let Err(e) = self.check_vote_valid(vote_msg) {
+            return e;
+        }
+        let vote_aggr_res = self.aggregate_qc(vote_msg, min_votes_for_qc);
+        if let VoteReceptionResult::NewQuorumCertificate(_) = vote_aggr_res {
+            return vote_aggr_res;
+        }
+        match self.aggregate_tc(vote_msg, min_votes_for_qc) {
+            Some(VoteReceptionResult::NewTimeoutCertificate(tc)) => {
+                VoteReceptionResult::NewTimeoutCertificate(tc)
+            }
+            _ => vote_aggr_res,
+        }
+    }
+
+    /// Check whether the newly inserted vote completes a QC
+    fn aggregate_qc(&mut self, vote_msg: &VoteMsg, min_votes_for_qc: usize) -> VoteReceptionResult {
+        // Note that the digest covers the ledger info information, which is also indirectly
+        // covering vote data hash (in its `consensus_data_hash` field).
+        let li_digest = vote_msg.ledger_info().hash();
+        let li_with_sig = self.li_digest_to_votes.entry(li_digest).or_insert_with(|| {
+            LedgerInfoWithSignatures::new(vote_msg.ledger_info().clone(), HashMap::new())
+        });
+        vote_msg
+            .signature()
+            .clone()
+            .add_to_li(vote_msg.author(), li_with_sig);
+
+        let num_votes = li_with_sig.signatures().len();
+        if num_votes >= min_votes_for_qc {
+            let quorum_cert = QuorumCert::new(vote_msg.vote_data().clone(), li_with_sig.clone());
+            return VoteReceptionResult::NewQuorumCertificate(Arc::new(quorum_cert));
+        }
+        VoteReceptionResult::VoteAdded(num_votes)
+    }
+
+    /// In case a timeout certificate is formed (there are 2f+1 votes in the same round) return the
+    /// new TimeoutCertificate, otherwise, return None.
+    fn aggregate_tc(
+        &mut self,
+        vote_msg: &VoteMsg,
+        min_votes_for_tc: usize,
+    ) -> Option<VoteReceptionResult> {
+        let round_signature = match vote_msg.round_signature() {
+            Some(round_signature) => round_signature.clone(),
+            None => {
+                return None;
+            }
+        };
+        let round = vote_msg.vote_data().block_round();
+        let author = vote_msg.author();
+        let tc = self
+            .round_to_tc
+            .entry(round)
+            .or_insert_with(|| TimeoutCertificate::new(round, HashMap::new()));
+        tc.add_signature(author, round_signature);
+        let num_votes = tc.signatures().len();
+        if num_votes >= min_votes_for_tc {
```

Comment:
Nit: 112 is only used once as well.

```suggestion
        if tc.signatures().len() >= min_votes_for_tc {
```